lib/ts/Extendible.h

```diff
@@ -0,0 +1,450 @@
+/** @file
```

Comment:
If the point here is to have multiple simultaneous readers, but just one writer with no simultaneous readers, wouldn't this work?

https://godbolt.org/g/vmeEFM

Reply:
If that where the point. I suppose I should document this better. 
The goal is to have readers access existing data while a writer is pending or actively updating a new copy of the data. Readers are only blocked while the pointer to the new data is updated.
Secondly I wanted to enforce scope locks of the writing code, so we could do something like
`write_ptr<vector<int>>(get_data_reader()).push(more_elements());` Where this locks the write lock, copies the list, appends more_elements. And because it immediately leaves scope, commits the changes, locks the read lock, updates the read pointer, and releases both locks.

Reply:
Not sure I understand.  Seem like you could perhaps just use std::shared_ptr, like:

std::shared_ptr<MyDB> current;

as a static storage variable.  The writer would allocate a new MyDB, copied from the current one, and changes, and update 'current'.  Readers would copy and use 'current'.  Old versions of MyDB would be deleted as readers stopped using them (changed or nulled their share_ptr instance pointing to them).

Reply:
Well we can't stop or restart a reader (a route request) every time a variable changes. This solution is referred to as "snapshot" or "copy-on-write", where you read the last available data, and don't care if it is updated after that point. Which in-turn means we can't delete the data while a reader is still using it.   

Imagine we store a list of IPs per host. A reader wants to iterate the list. A writer wants to replace the list. One of them has to copy the list. We chose the writer to copy the list because it should happen at a much lower frequency. Thus when we commit a writer, into the reader, that data becomes immutable, and will exist, unchanged until all references are deleted.

Reply:
I don't see how what you're saying here is inconsistent with my suggestion of using standard shared_ptr.

Reply:
https://stackoverflow.com/questions/14482830/stdshared-ptr-thread-safety

Reply:
After thinking about it, I don't see a really good approach using shared pointer.  I think maybe the best approach is to ping-pong between two buffers, one for shared reading, and the other locked by the writer thread:
```
#include <mutex>
#include <thread>
#include <atomic>

class OneWriterLock // but multiple simultaneous readers are allowed when not writing.
{
// Note:  the writer threads cannot be lower priority than reader threads.
private:
    std::mutex _countLock, _dataLock;
    int _readerCount = 0;

    // Include logic to prevent high read volume from locking out pending writes.
    // Use an atomic counter rather than a flag in case writes are done from mutliple
    // threads.
    std::atomic<int> _pendingWrite;

public:
    OneWriterLock() : _pendingWrite(0) {}

    void writeLock()
    {
        ++_pendingWrite;
        _dataLock.lock();
        --_pendingWrite;
    }

    void writeUnlock() { _dataLock.unlock(); }

    void readLock()
    {
        for ( ; ; )
        {
            _countLock.lock();
            
            if (_pendingWrite == 0) {
                break;
            }

            // Don't continue while there is a pending write, give other readers a chance
            // to finish so the write can proceed.
            _countLock.unlock();
            std::this_thread::yield();
        }
                
        ++_readerCount;
        if (_readerCount == 1) {
            _dataLock.lock();
        }
        _countLock.unlock();
    }

    void readUnlock()
    {
        _countLock.lock();
        --_readerCount;
        if (_readerCount == 0) {
            _dataLock.unlock();
        }
        _countLock.unlock();
    }

    // No copying.
    OneWriterLock(const OneWriterLock &) = delete;
    OneWriterLock & operator = (const OneWriterLock &) = delete;

    class ReadGuard
    {
    private:
        OneWriterLock &_lock;
    public:
        ReadGuard(OneWriterLock &lock) : _lock(lock)
        {
            _lock.readLock();
        }
        ~ReadGuard() { _lock.readUnlock(); }

      // No copying.
      ReadGuard(const ReadGuard &) = delete;
      ReadGuard & operator = (const ReadGuard &) = delete;
    };

    class WriteGuard
    {
    private:
        OneWriterLock &_lock;
    public:
        WriteGuard(OneWriterLock &lock) : _lock(lock)
        {
            _lock.writeLock();
        }
        ~WriteGuard() { _lock.writeUnlock(); }

      // No copying.
      WriteGuard(const WriteGuard &) = delete;
      WriteGuard & operator = (const WriteGuard &) = delete;
    };
};

/*
STATE TRANSITIONS

State is (count lock, data lock, reader count)

(N, N, 0)
writer takes data lock -> (N, Y, 0)
reader takes count lock -> (Y, N, 0)

(N, Y, 0)
writer releases data lock -> (N, N, 0)
reader takes count lock -> (Y, Y, 0)

(N, Y, 1)
reader takes count lock -> (Y, Y, 1)

(N, Y, >1)
reader takes count lock -> (Y, Y, >1)

(Y, N, 0)
writer takes data lock -> (Y, Y, 0)
reader increments reader count -> (Y, N, 1)
reader releases count lock -> (N, N, 0)

(Y, N, 1)
writer takes data lock -> (Y, Y, 1)
reader takes data lock -> (Y, Y, 1)

(Y, Y, 0)
writer releases data lock -> (Y, N, 0)
reader releases data lock -> (Y, N, 0)

(Y, Y, 1)
writer releases data lock -> (Y, N, 1)
reader releases count lock -> (N, Y, 1)
reader increments reader count -> (Y, Y, >1)
reader decrements reader count -> (Y, Y, 0)

(Y, Y, >1)
reader releases count lock -> (N, Y, >1)
reader increments reader count -> (Y, Y, >1)
reader decrements reader count -> (Y, Y, 1) or (Y, Y, >1)

*/

// This only works for one writer thread, or if the PingPong instance has an additional mutex to protect it.
template <class Data>
class PingPong
{
private:

    Data buf[2];

    OneWriterLock owl[2];

    std::atomic<int> select{0};

public:

    PingPong() {}

    // No copying.
    PingPong(const PingPong &) = delete;
    PingPong & operator = (const PingPong &) = delete;

    void init() { owl[1].writeLock(); }

    class Reader;

    friend class Reader;

    class Reader
    {
    friend class PingPong;

    private:
        Data *buf;
        OneWriterLock *owl;

        Reader() {}

    public:

        const Data *data() const { return buf; }

        // No copying.
        Reader(const Reader &) = delete;
        Reader & operator = (const Reader &) = delete;

        ~Reader()
        {
            if (owl)
            {
                owl->readUnlock();
            }
        }

        Reader(Reader &&that)
        {
            owl = that.owl;
            that.owl = nullptr;
            buf = that.buf;
            owl = nullptr;
        }

        Reader & operator = (Reader &&that)
        {
            if (this != &that)
            {
                owl = that.owl;
                owl = nullptr;
                buf = that.buf;
                owl = nullptr;
            }
        }
    };

    Reader reader()
    {
        int s = select;
        owl[s].readLock();
        Reader r;
        r.buf = buf + s;
        r.owl = owl + s;
        return r;
    }

    Data *write() { return buf + !select; }

    void swap()
    {
        int s = select;
        select = !select;
        owl[select].writeUnlock();
        owl[s].writeLock();
        buf[s] = buf[select];
    }
};

// Code to test compilation.

PingPong<int> pp;

void bar(const int *);
void bar2(int *);

void foo()
{
  {
      PingPong<int>::Reader r = pp.reader();
      bar(r.data());
  }
  bar2(pp.write());
  pp.swap();
}
```